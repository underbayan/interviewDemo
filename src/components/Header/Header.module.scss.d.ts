export const actionBarStyle: string;
export const brandStyle: string;
export const langStyle: string;
export const navStyle: string;
export const themeStyle: string;
