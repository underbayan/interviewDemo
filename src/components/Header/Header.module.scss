@import "@demo/designSystem";

.navStyle {
  width: 100vw;
  border-bottom: 1px solid var(--grays-5-color);
  position: fixed;
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  background-color: var(--grays-7-color);

  .langStyle {
    align-self: center;
    cursor: pointer;
    margin-right: 10px;
    text-transform: uppercase;
    padding: 6px 8px;
    border: 1px solid var(--grays-6-color);
    font-size: 0.75rem;
    height: 26px;
    box-sizing: border-box;

    &:hover {
      background-color: var(--grays-1-color);
      border-color: var(--grays-1-color);
    }
  }

  .actionBarStyle {
    justify-content: end
  }

  .themeStyle {
    @extend .langStyle;
    background-color: var(--grays-7-color);

  }

  .brandStyle {
    flex: 1;
    font-size: 14px;
    font-weight: 500;
    text-transform: uppercase;
  }
}