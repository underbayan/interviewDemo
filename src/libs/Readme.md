"Libs" refer to the libraries or modules that contain code that is meant to be reusable across different parts of an application. In the context of the specified constraint, the code in "libs" should not contain any mutations logic. Rather, it should be focused on providing functionalities that are shared across different parts of the application without altering any data. For instance, "libs" could contain helper functions, utility classes, or any other non-mutating code that can be used in different parts of the application.