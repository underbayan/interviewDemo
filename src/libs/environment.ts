import stringsEn from "../mock.strings.en.json"
import stringsZh from "../mock.strings.zh.json"

export const environment = {
  app: "demo",
  page: "/",
  lang: "en",
  version: "0.0.1",
  flights: "", // AB test
  config: {},  // Config for page
  strings: stringsEn,
  langMap: {
    en: stringsEn,
    zh: stringsZh
  }
  // TODO: i18n.detectLanguage() to get zh/en/fr lang config json
}