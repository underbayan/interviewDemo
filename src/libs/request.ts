export interface FetchDataType {
  [key: string]: any
}

export interface AbortableFetchResponse extends Response {
  data?: FetchDataType | string | ArrayBuffer | Blob
  message: string
}

export interface AbortableFetchOptions extends RequestInit {
  abort?: AbortSignal
}

export interface AbortableFetchPromise extends Promise<AbortableFetchResponse> {
  [AbortableKey](): void
}

export const AbortableKey = Symbol.for('abortableKey')
export const abortableFetch = (
  url: string,
  options?: AbortableFetchOptions
): AbortableFetchPromise => {
  const { abort, ...fetchOptions } = options ?? {}
  const ab = new AbortController()
  const rt = (fetch(url, { ...fetchOptions, signal: abort ? abort : ab.signal }) as Promise<Response>)
    .then(async (response) => {
      const contentType = response.headers.get('content-type')
      let data: FetchDataType | undefined | string

      if (contentType?.includes('application/json')) {
        data = await response.json()
      } else if (contentType?.includes('application/text')) {
        data = await response.text()
      } else if (contentType?.includes('application/opaque')) {
        data = await response.blob()
      } else {
        data = await response.arrayBuffer()
      }

      const result: AbortableFetchResponse = {
        ...response,
        status: response.status,
        message: response.statusText,
        data,
      }
      return result
    })
    .catch((e) => console.error(e)) as AbortableFetchPromise

  Object.defineProperty(rt, AbortableKey, { value: () => ab.abort(), writable: false })

  return rt
}
