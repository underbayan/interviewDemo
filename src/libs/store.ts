import { createSlice, configureStore } from '@reduxjs/toolkit'
import { useSelector, useDispatch } from 'react-redux'
import { environment } from "./environment"
import { Dispatch, AnyAction } from 'redux'
export type RootState = ReturnType<typeof store.getState>

const envSlice = createSlice({
  name: 'env',
  initialState: environment,
  reducers: {

    updateEnv(state, action) {

      const newState = { ...state, ...action?.payload }
      if (action?.payload?.lang) newState.strings = newState.langMap[action.payload?.lang]
      return newState
    }
  }
})
export const store = configureStore({
  reducer: {
    env: envSlice.reducer
  },
})
interface EnvUpdatePayload {
  lang?: string
  [key: string]: any
}




export const useI18n = () => useSelector((state: RootState) => state.env.strings)
export const useEnv = () => useSelector((state: RootState) => state.env)
export const useUpdateEnv = (): ((env: EnvUpdatePayload) => void) => {
  const dispatch = useDispatch()

  const updateEnvAsync = (env: EnvUpdatePayload) => (dispatch: Dispatch) => {
    setTimeout(() => { dispatch(envSlice.actions.updateEnv(env)) }, 15)
  }

  return (env: EnvUpdatePayload) => dispatch(updateEnvAsync(env) as unknown as AnyAction)
}


