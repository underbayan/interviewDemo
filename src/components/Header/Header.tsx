import { navStyle, brandStyle, langStyle, themeStyle, actionBarStyle } from "./Header.module.scss"
import { useI18n, useUpdateEnv, useEnv } from "../../libs/store"
import { memo, useCallback } from "react"
import {
  MouseEventHandler
} from 'react';
type StrType = Record<string, string>
interface HeaderViewProps {
  str: StrType
  changeLang: MouseEventHandler<HTMLDivElement>
  nextLang: string
  changeTheme: MouseEventHandler<HTMLDivElement>
}
const HeaderView = memo(({ str, changeLang, nextLang, changeTheme }: HeaderViewProps) => <header className={navStyle}>
  <div className={"row " + actionBarStyle}>
    <div className={"p-xs-1 p-sm-2 " + brandStyle}> {str.brand} </div>
    <div className={langStyle} onClick={changeLang}> {nextLang}</div>
    <div className={themeStyle} onClick={changeTheme}> ðŸŒ“ </div>
  </div>
</header>)

export const Header = memo(() => {
  const str: StrType = useI18n()
  const updateEnv = useUpdateEnv()
  const env = useEnv()

  const nextLang = env.lang === "en" ? "zh" : "en"
  const changeTheme = useCallback(() => document.getElementsByTagName('body')[0].classList.toggle("dark"), [])

  return (<HeaderView
    str={str}
    nextLang={nextLang}
    changeLang={updateEnv.bind(null, { lang: nextLang })}
    changeTheme={changeTheme}
  />
  )
})




