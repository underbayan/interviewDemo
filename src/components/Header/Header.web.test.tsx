import React from 'react';
import { render } from '@testing-library/react';
import { test, expect } from 'vitest';
import { Header } from './Header';
import { store } from "@demo/libs/store"
import { Provider } from 'react-redux'

test('Header renders correctly', () => {
  const { container } = render(<Provider store={store}>
    <Header>Click me</Header>
  </Provider>);
  expect(container).toMatchSnapshot();
});
