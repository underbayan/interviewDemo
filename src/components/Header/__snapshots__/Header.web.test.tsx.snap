// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Header renders correctly 1`] = `
<div>
  <header>
    <div
      class="row undefined"
    >
      <div
        class="p-xs-1 p-sm-2 undefined"
      >
         
        Broccoli & co.
         
      </div>
      <div>
         
        zh
      </div>
      <div>
         ðŸŒ“ 
      </div>
    </div>
  </header>
</div>
`;
